docs/multilingual_tokenizer_documentation (#357)
cisco/mindmeld: source/userguide/text_preparation_pipeline.rst

We see that the original text is split at the character level for non-latin characters but the number "365" remains as an unsegmented token.

.. code:: python

    ['1', '年', 'は', '365', '日']

Spacy Tokenizer
^^^^^^^^^^^^^^^
The :attr:`SpacyTokenizer` splits up a sentence using `Spacy's language models <https://spacy.io/models>`_.
Supported languages include English (en), Spanish (es), French (fr), German (de), Danish (da), Greek (el), Portuguese (pt), Lithuanian (lt), Norwegian Bokmal (nb), Romanian (ro), Polish (pl), Italian (it), Japanese (ja), Chinese (zh), Dutch (nl).
If the required Spacy model is not already present it will automatically downloaded during runtime. 
Let's use the :attr:`SpacyTokenizer` to tokenize the Japanese translation of "The gentleman is gone, no one knows why it happened!": 

.. code:: python

    from mindmeld.text_preparation.tokenizers import SpacyTokenizer
    
    sentence_ja = "紳士が過ぎ去った、 なぜそれが起こったのか誰にも分かりません！"
    spacy_tokenizer_ja = SpacyTokenizer(language="ja", spacy_model_size="lg")
    tokens = spacy_tokenizer_ja.tokenize(sentence_ja)

We see that the original text is split semantically and not simply by whitespace.

.. code:: python

>     ['紳士', 'が', '過ぎ', '去っ', 'た', '、', 'なぜ', 'それ', 'が', '起こっ', 'た', 'の', 'か', '誰', 'に', 'も', '分かり', 'ませ', 'ん', '！']

Reviewer: probably clearer to reuse "背の高い男性" here

^^^^^^^^^^^^^^^^^^^^^^^^^^^
This section includes boilerplate code to build a custom tokenizer class. Let's rebuild a :attr:`CharacterTokenizer` class that creates a token for each character in a string as long as the
character is not a space. A custom tokenizer must extend from MindMeld's abstract :attr:`Tokenizer` class:

.. code:: python

    from abc import ABC, abstractmethod

    class Tokenizer(ABC):
        """Abstract Tokenizer Base Class."""

        @abstractmethod
        def tokenize(self, text):
            """
            Args:
                text (str): The text to tokenize.
            Returns:
                tokens (List[Dict]): List of tokenized tokens which a represented as dictionaries.
                    Keys include "start" (token starting index), and "text" (token text).
                    For example: [{"start": 0, "text":"hello"}]
            """
            raise NotImplementedError("Subclasses must implement this method")

Note that any MindMeld tokenizer must return the final tokens as a list of dictionaries. Where each dictionary represents a single token and contains the "start" index of the token and the "text" of the token.

> Here is an example of the expected output for the tokens generated when tokenizing the phrase "Hi Andy": [{"start": 0, "text":"Hi"}, {"start": 3, "text":"Andy"}].

Reviewer: a point of clarity: if the preprocessor changes the string, are these indices referencing the raw text or the processed text?

Author: Good point, these indices reference the processed text. I'll add this information here

        @abstractmethod
        def normalize(self, text):
            """
            Args:
                text (str): Text to normalize.
            Returns:
                normalized_text (str): Normalized Text.
            """
            raise NotImplementedError("Subclasses must implement this method")

With this in mind, let's recreate MindMeld's :attr:`Lowercase` normalizer class.

.. code:: python

    from mindmeld.text_preparation.normalizers import Normalizer

    class Lowercase(Normalizer):

        def normalize(self, text):
            return text.lower()

This normalizer would transform the text "I Like to Run!" to "i like to run!".
The steps to use a custom Normalizer in your application are explained here.

.. note::

    MindMeld normalizes queries on a per-token basis. Custom normalizers should be designed to normalize individual tokens and not sentences as a whole.

Stemming
--------
Stemming is the process of reducing a word to its stem or root. If a stemmer is not specified in the :attr:`TEXT_PREPARATION_CONFIG`, then MindMeld will automatically select a stemmer
based on the language of the application.

EnglishNLTKStemmer
^^^^^^^^^^^^^^^^^^

> The :attr:`EnglishNLTKStemmer` stemmer uses the :attr:`PorterStemmer` from the nltk library. The porter stemmer implements a series of rules that determine common suffixes from sentences.

Reviewer: ```suggestion
The :attr:`EnglishNLTKStemmer` stemmer uses a modified version of the :attr:`PorterStemmer` from the nltk library. The Porter stemmer implements a series of rules that removes common suffixes, and this version of it removes inflectional suffixes but leaves (most) derivational suffixes in place.
```

> This includes removing the letter "s" from plural words or "ing" from gerunds.

Reviewer: ```suggestion
This includes removing the final letters "s"/"es" from plural words or "ing" from gerunds, but leaving more meaningful suffixes like "tion" and "ment" alone.
```

Let's take a look at a few examples of the :attr:`EnglishNLTKStemmer`. First we'll make an instance of the stemmer:

.. code:: python

    from mindmeld.text_preparation.stemmers import EnglishNLTKStemmer
    english_nltk_stemmer = EnglishNLTKStemmer()

> Now let's stem the words "running" and "pearls".

Reviewer: ```suggestion
Now let's stem the words "running" and "governments".
```

.. code:: python

    >>> print(english_nltk_stemmer.stem_word("running"))
    >>> run

>     >>> print(english_nltk_stemmer.stem_word("pearls"))

Reviewer: ```suggestion
    >>> print(english_nltk_stemmer.stem_word("governments"))
```

>     >>> pearl

Reviewer: ```suggestion
    >>> government
```

> As expected, the stemmer removes "ing" from "running" and the "s" from "pearls" to create stemmed words.

Reviewer: ```suggestion
As expected, the stemmer removes "ing" from "running" and the "s" from "governments" to create stemmed words.
```

SnowballNLTKStemmer
^^^^^^^^^^^^^^^^^^^

> The :attr:`SnowballNLTKStemmer` stemmer works in a similar manner to the :attr:`EnglishNLTKStemmer`, however, it offers support for a larger set of languages.

Reviewer: ```suggestion
The :attr:`SnowballNLTKStemmer` stemmer works in a similar manner to the :attr:`EnglishNLTKStemmer`, however, it removes more suffixes and offers support for a larger set of languages.
```

Namely, the :attr:`SnowballNLTKStemmer` supports Danish (da), Dutch (nl), Finnish (fi), French (fr), German (de), Hungarian (hu), Italian (it), Norwegian (nb), Portuguese (pt), Romanian (ro), Russian (ru), Spanish (es) and Swedish (sv).

> To create an instance of the :attr:`SnowballNLTKStemmer`, we can use MindMeld's :attr:`StemmerFactory`.

Reviewer: Why is this better than directly calling 'SnowballNLTKStemmer'? (I suspect the answer is that the stemmer factory lets you call by language code, where as the stemmer class expects a full language name, but is there more to it?) Notably, the full Snowball/Porter stemmer with derivational suffixes removed is also accessible through `SnowballNLTKStemmer("english")`.

Author: Yes, that's correct. The stemmer factory lets you call by language code and converts that code into a lower-cased version of the language like "spanish" for exampe. Aside from this, there isn't any other main difference. A user could alternatively do SnowballNLTKStemmer("spanish"), for example. Should I switch the example to use the latter?

.. code:: python

    from mindmeld.text_preparation.stemmers import StemmerFactory
    es_snowball_stemmer = StemmerFactory.get_stemmer_by_language("es")

> Now let's stem the words "corriendo" ("running") and "perlas" ("pearls").

Reviewer: ```suggestion
Now let's stem the words "corriendo" ("running") and "gobiernos" ("governments").
```

.. code:: python

    >>> print(english_nltk_stemmer.stem_word("corriendo"))
    >>> corr

>     >>> print(english_nltk_stemmer.stem_word("perlas"))

Reviewer: ```suggestion
    >>> print(english_nltk_stemmer.stem_word("gobiernos"))
```

>     >>> perl

Reviewer: ```suggestion
    >>> gobi
```

As expected, the stemmer removes "iendo" from "corriendo" and the "as" from "perlas" to create stemmed words.

Creating a Custom Stemmer
^^^^^^^^^^^^^^^^^^^^^^^^^
This section includes boilerplate code to build a custom stemmer class.
A custom stemmer must extend from MindMeld's abstract :attr:`Stemmer` class:

.. code:: python

    from abc import ABC, abstractmethod

    class Stemmer(ABC):

        @abstractmethod
        def stem_word(self, word):
            """
            Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.

            Args:
                word (str): The word to stem

            Returns:
                stemmed_word (str): A stemmed version of the word
            """
            raise NotImplementedError

> Let's create a stemmer that only removes the "-ing" suffix if found at the end of a word. We'll call it the :attr:`GerundSuffixStemmer`.

Reviewer: A more plausible example would be more helpful--would it be to much to show how to use spacy lexemes as stems?

.. code:: python

    from mindmeld.text_preparation.stemmers import Stemmer

    class GerundSuffixStemmer(Stemmer):

        def stem_word(self, word):
            """
            Stemmer that removes the "-ing" suffix if found at the end of a word.

            Args:
                word (str): The word to stem

            Returns:
                stemmed_word (str): A stemmed version of the word
            """
            if word.endswith("ing"):
                return word[:-len("ing")]
            return word

This stemmer would transform "jumping" to "jump".

> The steps to use a custom Stemmer in your application are explained here.

Reviewer: "below" or link?

Author: Added internal links to the appropriate section

Using a Custom TextPreparationPipeline for your Application
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
As a recap, every MindMeld project is also a Python package and has an ``__init.py__`` file at the root level.
This package also contains an *application container* -- a container for all of the logic and functionality for your application.
This application container enumerates all of the dialogue states and their associated handlers, and should be defined as ``app`` in the application's Python package.
To use a :attr:`TextPreparationPipeline` with custom components, we must pass in a custom object into the application container in ``__init.py__``.

> Let's first take a look at an example of a an ``__init.py__`` file before a custom :attr:`TextPreparationPipeline` used.

Reviewer: ```suggestion
Let's first take a look at an example of an ``__init.py__`` file before a custom :attr:`TextPreparationPipeline` used.
```

+-----------------------------------------------------+--------------------------------------------------------------------------------------------------+---------------------------------------+-------------------------------+
| ReplaceSpecialCharsBetweenDigitsAndLettersWithSpace | Replaces special characters between digits and letters with a space.                             | "Coding 24^^!#%%hours 7##%days"       | "Coding 24 hours 7 days"      |
+-----------------------------------------------------+--------------------------------------------------------------------------------------------------+---------------------------------------+-------------------------------+
| ReplaceSpecialCharsBetweenLettersWithSpace          | Replaces special characters between letters and letters with a space.                            | "Coding all^^!#%%hours seven##%days"  | "Coding all hours seven days" |
+-----------------------------------------------------+--------------------------------------------------------------------------------------------------+---------------------------------------+-------------------------------+

> The last 5 rules above remove special characters in different contexts. These special characters can be specified in the config using the key, :attr:`keep_special_chars`.

Reviewer 2: I think it would be better to name the 5 here since if we add more to the table, we have to also update it here.

Author: Sure, added it in.
